### PR TITLE
TINY-6280: Fix fullscreen plugin not working with multiple editors and shadowdom

### PR DESCRIPTION
**Pull request by ltrouton**
Related Ticket: TINY-6280

Description of Changes:
* Refactor `fullscreen.less` so that the selectors are compatible with both Body and ShadowDom
* Make sure `tox-fullscreen` is added to the shadow host and that it has the correct z-index
* Update test to run in Body and ShadowDom

Pre-checks:
* [X] Changelog entry added
* [X] Tests have been added (if applicable)
* [x] ~~Branch prefixed with `feature/` for new features (if applicable)~~
* [X] License headers added on new files (if applicable)

Review:
* [X] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

**Review by lostkeys — APPROVED** [review]

File: modules/tinymce/src/plugins/fullscreen/main/ts/core/Actions.ts [hunk 1]
```diff
@@ -112,14 +112,21 @@ const toggleFullscreen = (editor: Editor, fullscreenState: Cell<ScrollInfo | nul
   const iframe = editor.iframeElement;
   const iframeStyle = iframe.style;
 
+  const handleClasses = (handler: (elm: string | Node | Node[], cls: string) => void) => {
+    handler(body, 'tox-fullscreen');
+    handler(documentElement, 'tox-fullscreen');
+    handler(editorContainer, 'tox-fullscreen');
+    if (SugarShadowDom.isOpenShadowHost(fullscreenRoot)) {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/fullscreen/main/ts/core/Actions.ts:119]: This isn't the correct function to use, as this needs to work with a closed shadow root as well. You probably need to do something like this instead:
```
SugarShadowDom.getShadowRoot(SugarElement.fromDom(editorContainer)).map((root) => {
  const host = SugarShadowDom.getShadowHost(root);
  handler(host.dom, 'tox-fullscreen tox-shadowhost');
})
```

File: modules/tinymce/src/plugins/fullscreen/main/ts/core/Actions.ts [hunk 2]
```diff
@@ -112,14 +112,21 @@ const toggleFullscreen = (editor: Editor, fullscreenState: Cell<ScrollInfo | nul
...
+      handler(fullscreenRoot.dom, 'tox-fullscreen tox-shadowhost');
```
Comment (lnewson) [inline modules/tinymce/src/plugins/fullscreen/main/ts/core/Actions.ts:120]: This while technically works, it only happens by chance. This should be 2 separate calls, one for each class.

File: modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts [hunk 3]
```diff
@@ -33,78 +47,108 @@ UnitTest.asynctest('browser.tinymce.plugins.fullscreen.FullScreenPluginTest', (s
...
+        ])
+      ]),
+      Log.chainsAsStep('TBA', 'FullScreen: Toggle fullscreen and remove editor should clean up classes', [
+        Chain.inject(editor),
+        Chain.fromParent(Chain.identity, [
+          Chain.op((editor: Editor) => editor.execCommand('mceFullScreen', true)),
+          cAssertApiAndLastEvent('After fullscreen command', true),
+          cAssertPageState('After fullscreen command', true)
+        ]),
+        McEditor.cRemove,
```
Comment (lnewson) [inline modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts:139]: This shouldn't be here, as you've not constructed an editor using `McEditor`. If you're switching to using TinyLoader, then we'll need to find another way to do this.

**Review by lnewson — CHANGES_REQUESTED** [review]
Looks good overall Luke, just marking as requested changes because of a few bugs.

File: modules/tinymce/src/plugins/fullscreen/main/ts/core/Actions.ts (hunk 1 above)
Comment (ltrouton) [reply modules/tinymce/src/plugins/fullscreen/main/ts/core/Actions.ts]: Thanks for the suggestion. This should be fixed now.

File: modules/tinymce/src/plugins/fullscreen/main/ts/core/Actions.ts (hunk 2 above)
Comment (ltrouton) [reply modules/tinymce/src/plugins/fullscreen/main/ts/core/Actions.ts]: Fixed

File: modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts [hunk 4]
```diff
@@ -1,17 +1,31 @@
...
+    Assertions.assertEq('No windows exist', 0, dialogs().length);
+  }),
+  Guard.addLogging('Close window')
+);
+
+const cWaitForDialog = (ariaLabel: string) =>
   Chain.control(
-    Chain.fromChainsWith(SugarBody.body(), [
+    Chain.fromChains([
+      Chain.mapper(getContentContainer),
```
Comment (metricjs) [inline modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts:28]: If I'm reading this right, this is dependent on running after a Chain that happens to pass along `editor`? I'm wondering if it would be safer/easier to read if we inject `editor` in case... Not sure though.

File: modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts [hunk 5]
```diff
@@ -33,78 +47,108 @@ UnitTest.asynctest('browser.tinymce.plugins.fullscreen.FullScreenPluginTest', (s
...
-      editor.windowManager.close();
-      Assertions.assertEq('No windows exist', 0, dialogs().length);
-    }),
-    Guard.addLogging('Close window')
-  );
+  const cAsssertEditorContainerAndSinkState = (label: string, shouldExist: boolean) =>
+    Chain.label(
+      `${label}: Editor container and sink should ${shouldExist ? '' : 'not'} have "tox-fullscreen" class and z-index`,
+      Chain.fromChains([
+        Chain.fromIsolatedChains([
```
Comment (metricjs) [inline modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts:76]: It _is_ 7PM on a Friday, but I'm finding this rather hard to follow. I'm wondering if NamedChains - much as I dislike using them - might be more readable?

File: modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts (hunk 3 above)
Comment (metricjs) [reply modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts:139]: Looks like it just wasn't deleted when changing the tests?

Comment (ltrouton) [reply modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts]: I have restructured the code so this can be kept to allow the class removal logic on editor removal to be tested.

File: modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts [hunk 6]
```diff
@@ -1,72 +1,134 @@
...
+      })
+    )
+  ])
+);
+
+const cWaitForDialog = (ariaLabel: string) =>
+  Chain.label(
+    'Looking for dialog with an aria-label: ' + ariaLabel,
+    Chain.fromChains([
+      NamedChain.direct('editor', Chain.mapper(getContentContainer), 'contentContainer'),
```
Comment (lnewson) [inline modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts:39]: This isn't a place named chains should be used as the items aren't used out of order or multiple times. Also where is `editor` coming from as that's never assigned and there's no `NamedChain.asChain` call? I suspect you're attempting to use something that just happens to work but shouldn't.

Anyways, this could be:
```
Chain.mapper(getContentContainer),
UiFinder.cWaitFor('Waiting for dialog', '[role="dialog"]'),
Chain.op((dialog) => { ... }
```

File: modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts [hunk 7]
```diff
@@ -1,72 +1,134 @@
...
+      )
+    ])
+  );
+
+const cAssertHtmlAndBodyState = (label: string, shouldExist: boolean) => {
+  const selector = shouldExist ? 'root:.tox-fullscreen' : 'root::not(.tox-fullscreen)';
+  return Chain.label(
+    `${label}: Body and Html should ${shouldExist ? '' : 'not'} have "tox-fullscreen" class`,
+    Chain.fromChains([
+      NamedChain.writeValue('docBody', SugarBody.body()),
```
Comment (lnewson) [inline modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts:60]: Another place we shouldn't be using named chains.

**Review by lnewson — CHANGES_REQUESTED** [review]
Marking as changes needed, as NamedChains appear to be used rather incorrectly here sorry @ltrouton.

File: modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts (hunk 6 above)
Comment (lnewson) [reply modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts:39]: To clarify more, you're mixing NamedChains with Chains. You're also trying to pass the named chain state across multiple functions, which makes it **very** hard to follow and messy, as you never know what state you have.

File: modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts [hunk 8]
```diff
@@ -1,17 +1,38 @@
...
+  'Close window',
+  Chain.op((editor: Editor) => {
+    const dialogs = () => UiFinder.findAllIn(getContentContainer(editor), '[role="dialog"]');
+    Assertions.assertEq('One window exists', 1, dialogs().length);
+    editor.windowManager.close();
+    Assertions.assertEq('No windows exist', 0, dialogs().length);
+  })
+);
+
+const cWaitForDialog = (ariaLabel: string) =>
```
Comment (lnewson) [inline modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts:31]: Maybe add the return type here to make it clear this needs an editor? eg `(ariaLabel: string): Chain<Editor, Editor> => `

File: modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts [hunk 9]
```diff
@@ -22,51 +43,86 @@ export const cWaitForDialog = (ariaLabel) =>
...
+    Chain.fromIsolatedChains([
+      Chain.inject(SugarBody.body()),
+      UiFinder.cFindIn(selector),
+      Chain.inject(SugarElement.fromDom(document.documentElement)),
+      UiFinder.cFindIn(selector)
+    ])
+  );
+};
+
+const cAsssertEditorContainerAndSinkState = (label: string, shouldExist: boolean) =>
```
Comment (lnewson) [inline modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts:62]: Another place we should maybe add the return type?

File: modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts [hunk 10]
```diff
@@ -76,35 +132,68 @@ UnitTest.asynctest('browser.tinymce.plugins.fullscreen.FullScreenPluginTest', (s
...
+    setupEditor: () => {
+      const shadowHost = SugarElement.fromTag('div');
+      Insert.append(SugarBody.body(), shadowHost);
+      const sr = SugarElement.fromDom(shadowHost.dom.attachShadow({ mode: 'open' }));
+      const editorDiv = SugarElement.fromTag('div');
+      Insert.append(sr, editorDiv);
+      return McEditor.cFromElement(editorDiv, settings);
+    },
+    cleanupEditor: () => Chain.fromChains([
+      Chain.op((editor: Editor) => {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts:154]: This is correct? I think this needs to happen in the opposite order, otherwise the editor will be removed from the DOM before it's shutdown properly.

**Review by lnewson — APPROVED** [review]
Looks much better, thanks @ltrouton!

File: modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts (hunk 10 above)
Comment (ltrouton) [reply modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts:154]: Thanks for picking this up. It should be fixed now.

**Review by metricjs — APPROVED** [review]

File: modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts [hunk 11]
```diff
@@ -22,51 +43,87 @@ export const cWaitForDialog = (ariaLabel) =>
...
+      UiFinder.cFindIn(selector),
+      Chain.inject(SugarElement.fromDom(document.documentElement)),
+      UiFinder.cFindIn(selector)
+    ])
+  );
+};
+
+const cAsssertEditorContainerAndSinkState = (label: string, shouldExist: boolean): Chain<Editor, Editor> =>
+  Chain.label(
+    `${label}: Editor container and sink should ${shouldExist ? '' : 'not'} have "tox-fullscreen" class and z-index`,
```
Comment (JamesToohey) [inline modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts:64]: An _ultra_ minor thing, but this will result in a double space, maybe:

```suggestion
    `${label}: Editor container and sink should ${shouldExist ? 'have' : 'not have'} "tox-fullscreen" class and z-index`,
```

or


```suggestion
    `${label}: Editor container and sink should${shouldExist ? ' ' : ' not '}have "tox-fullscreen" class and z-index`,
```

**Review by JamesToohey — APPROVED** [review]

File: modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts (hunk 11 above)
Comment (JamesToohey) [reply modules/tinymce/src/plugins/fullscreen/test/ts/browser/FullScreenPluginTest.ts:64]: I don't mind if this isn't changed, just noticed it
